ApiDoctor.Console/Program.cs

```diff
@@ -1960,14 +1957,149 @@ private static async Task<bool> GenerateSnippetsAsync(GenerateSnippetsOptions op
+            var fileFullPath = Path.Combine(snippetsPath, fileName);
+
+            FancyConsole.WriteLine(FancyConsole.ConsoleSuccessColor, $"Reading {fileFullPath}");
+
+            string codeSnippet = null;
+            if (File.Exists(fileFullPath))
+                codeSnippet = File.ReadAllText(fileFullPath);
+            else
+                FancyConsole.WriteLine(FancyConsole.ConsoleErrorColor, $"Error: file '{fileName}' does not exist");
+
+            return codeSnippet;
+        }
+
+        private static Dictionary<string, HashSet<string>> SnippetsLanguageSetForDocument = new Dictionary<string, HashSet<string>>();
+
+        private static HashSet<string> GetSnippetsLanguagesToIncludeForMethod(MethodDefinition sourceMethod, MethodDefinition[] methodCollection, string[] languageOptions, string snippetsPath, string snippetPrefix)
+        {
+            var languages = LookupGeneratedSnippetsLanguagesBySnippetPrefix(snippetPrefix, snippetsPath);
+            if (!languages.Any())
+            {
+                FancyConsole.WriteLine(FancyConsole.ConsoleWarningColor, $"Failed to fetch snippets for method: {sourceMethod.Identifier}");
+                return new HashSet<string>();
+            }
+            else if (SnippetsLanguageSetForDocument.TryGetValue(sourceMethod.SourceFile.DisplayName, out var storedLanguages))
+            {
+                return storedLanguages;
+            }
+            else
+            {
+                languages ??= new HashSet<string>();
+                var methodsInSameFileAsCurrentMethod = methodCollection
+                    .Where(x => x.SourceFile.DisplayName == sourceMethod.SourceFile.DisplayName &&
+                        x.Identifier != sourceMethod.Identifier)
+                    .ToList();
+                foreach (var method in methodsInSameFileAsCurrentMethod)
+                {
+                    // no need to continue if we already have the full list of languages
+                    if (languages.Count == languageOptions.Length)
+                        break;
+
+                    var snippetLanguagesForMethod = LookupGeneratedSnippetsLanguagesBySnippetPrefix(snippetPrefix, snippetsPath);
+                    languages.UnionWith(snippetLanguagesForMethod);
+                }
+
+                // just so that we have the correct casing displayed as tab names in docs e.g. PowerShell and not powershell
+                languages = languageOptions.Where(x => languages.Contains(x, StringComparer.OrdinalIgnoreCase)).ToHashSet();
+
+                SnippetsLanguageSetForDocument.Add(sourceMethod.SourceFile.DisplayName, languages);
+                return languages;
+            }
+        }
+
+        private static HashSet<string> LookupGeneratedSnippetsLanguagesBySnippetPrefix(string snippetPrefix, string snippetsPath)
+        {
+            var expectedFileName = $"{Path.Combine(snippetsPath, snippetPrefix)}---";
+            return Directory.EnumerateFiles(snippetsPath)
```

Comment:
Since the parent function is called pretty frequently, it may be better to call `Directory.EnumerateFiles(snippetsPath)` once and load the collection to memory and perform the file name filtering on the collection to reduce IO ops.